core/store/src/trie/trie_storage.rs

```diff
@@ -263,22 +263,23 @@ pub trait TrieStorage {
 pub struct TrieRecordingStorage {
     pub(crate) store: Store,
     pub(crate) shard_uid: ShardUId,
-    pub(crate) recorded: RefCell<HashMap<CryptoHash, Vec<u8>>>,
+    pub(crate) recorded: RefCell<HashMap<CryptoHash, Arc<[u8]>>>,
 }
 
 impl TrieStorage for TrieRecordingStorage {
     fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Arc<[u8]>, StorageError> {
         if let Some(val) = self.recorded.borrow().get(hash) {
-            return Ok(val.as_slice().into());
+            return Ok(val.clone());
         }
         let key = TrieCachingStorage::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
         let val = self
             .store
             .get(DBCol::State, key.as_ref())
             .map_err(|_| StorageError::StorageInternalError)?;
         if let Some(val) = val {
-            self.recorded.borrow_mut().insert(*hash, val.clone());
-            Ok(val.into())
+            let val = Arc::from(val);
```

Comment:
One think I was wondering that maybe db should be returning an `Arc` already? In theory, that'd allow us saving on one memcpy. On the other hand, it's unclear what we want here in an ideal world: an `Arc<[u8]>`, a `Bytes` (so that slicing is cheap) or `Arc<T: BorshSerialize>`.

Reply:
Yes, that’s a good point. We’re currently doing RocksDB’s internal slice representation → Vec → Arc. Returning RocksDB’s type in Database interface would probably not be kosher but we could offer two sets of methods: one returning Vec and another returning Arc.

Reply:
We could do something like 

```
pub fn with_value(&self, key: &[u8], f: impl FnOnce(Option<&[u8]>))
```

Reply:
Can you have generic in traits?

Reply:
In traits yes, in trait objects no.  So this would have to be `&dyn Fn` then, or two methods (but then you won't be able to deserialize without a copy).